pkg/rbd/rbd_journal.go

```diff
@@ -0,0 +1,307 @@
+	}
+
+	key, err := getKey(rbdSnap.AdminID, credentials)
+	if err != nil {
+		return false, err
+	}
+
+	snapUUID, err := snapJournal.CheckReservation(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+		rbdSnap.RequestName, rbdSnap.RbdImageName)
+	if err != nil {
+		return false, err
+	}
+	if snapUUID == "" {
+		return false, nil
+	}
+	rbdSnap.RbdSnapName = snapJournal.NamingPrefix() + snapUUID
+
+	// Fetch on-disk image attributes
+	err = updateSnapWithImageInfo(rbdSnap, credentials)
+	if err != nil {
+		if _, ok := err.(ErrSnapNotFound); ok {
+			err = snapJournal.UndoReservation(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+				rbdSnap.RbdSnapName, rbdSnap.RequestName)
+			return false, err
+		}
+		return false, err
+	}
+
+	// found a snapshot already available, process and return its information
+	rbdSnap.SnapID, err = util.GenerateVolID(rbdSnap.Monitors, rbdSnap.AdminID, key, rbdSnap.Pool,
+		rbdSnap.ClusterID, snapUUID, volIDVersion)
+	if err != nil {
+		return false, err
+	}
+
+	klog.V(4).Infof("Found existing snap (%s) with snap name (%s) for request (%s)",
+		rbdSnap.SnapID, rbdSnap.RbdSnapName, rbdSnap.RequestName)
+
+	return true, nil
+}
+
+/*
+Check comment on checkSnapExists, to understand how this function behaves
+
+**NOTE:** These functions manipulate the rados omaps that hold information regarding
+volume names as requested by the CSI drivers. Hence, these need to be invoked only when the
+respective CSI snapshot or volume name based locks are held, as otherwise racy access to these
+omaps may end up leaving the omaps in an inconsistent state.
+*/
+func checkVolExists(rbdVol *rbdVolume, credentials map[string]string) (bool, error) {
+	err := validateRbdVol(rbdVol)
+	if err != nil {
+		return false, err
+	}
+
+	key, err := getKey(rbdVol.AdminID, credentials)
+	if err != nil {
+		return false, err
+	}
+
+	imageUUID, err := volJournal.CheckReservation(rbdVol.Monitors, rbdVol.AdminID, key, rbdVol.Pool,
+		rbdVol.RequestName, "")
+	if err != nil {
+		return false, err
+	}
+	if imageUUID == "" {
+		return false, nil
+	}
+	rbdVol.RbdImageName = volJournal.NamingPrefix() + imageUUID
+
+	// NOTE: Return volsize should be on-disk volsize, not request vol size, so
+	// save it for size checks before fetching image data
+	requestSize := rbdVol.VolSize
+	// Fetch on-disk image attributes and compare against request
+	err = updateVolWithImageInfo(rbdVol, credentials)
+	if err != nil {
+		if _, ok := err.(ErrImageNotFound); ok {
+			err = volJournal.UndoReservation(rbdVol.Monitors, rbdVol.AdminID, key, rbdVol.Pool,
+				rbdVol.RbdImageName, rbdVol.RequestName)
+			return false, err
+		}
+		return false, err
+	}
+
+	// size checks
+	if rbdVol.VolSize < requestSize {
+		err = fmt.Errorf("image with the same name (%s) but with different size already exists",
+			rbdVol.RbdImageName)
+		return false, ErrVolNameConflict{rbdVol.RbdImageName, err}
+	}
+	// TODO: We should also ensure image features and format is the same
```

Comment:
I added this TODO because, we check if an existing volume is of a different (smaller) size than the request and hence is an error of type ErrVolNameConflict (also translates to the CSI error AlreadyExists) is returned. This check needs to be extended to check if the image requested has the same image features and format, as that of the image found existing for this request name.

Reply:
Sure, I get that , but wondering what comes under `image features` @ShyamsundarR

Reply:
> Sure, I get that , but wondering what comes under `image features` @ShyamsundarR

Currently I would assume this is "layering" as mentioned in the [storage class](https://github.com/ceph/ceph-csi/blob/b6b7cf2c3d42f08b5200068dc6f73ff5d822cec4/examples/rbd/storageclass.yaml#L23-L25).